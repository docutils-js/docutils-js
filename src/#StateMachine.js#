class StateMachine {
    /* 
        Initialize a `StateMachine` object; add state objects.

       Parameters:

        - `state_classes`: a list of `State` (sub)classes.
        - `initial_state`: a string, the class name of the initial state.
        - `debug`: a boolean; produce verbose output if true (nonzero).
	*/

    constructor(stateClasses, initialState, debug) {
	if(debug === undefined) {
	    debug = false;
	}
	this.inputLines = undefined;
	this.inputOffset = 0;
	this.line = undefined;
	this.lineOffset = -1;
	this.debug = debug;
	this.initialState = initialState;
	this.currentState = initialState;
	this.states = {};
	this.addStates(stateClasses);
	this.observers = []
	//       self._stderr = ErrorOutput()
    }

    unlink() {
	Object.values(this.states).forEach(s => s.unlink());
	this.states = undefined;
    }

    run(inputLines, inputOffset = 0, context, inputSource, initialState) {
	/*
        Run the state machine on `input_lines`. Return results (a list).

        Reset `self.line_offset` and `self.current_state`. Run the
        beginning-of-file transition. Input one line at a time and check for a
        matching transition. If a match is found, call the transition method
        and possibly change the state. Store the context returned by the
        transition method to be passed on to the next transition matched.
        Accumulate the results returned by the transition methods in a list.
        Run the end-of-file transition. Finally, return the accumulated
        results.

        Parameters:

        - `input_lines`: a list of strings without newlines, or `StringList`.
        - `input_offset`: the line offset of `input_lines` from the beginning
          of the file.
        - `context`: application-specific storage.
        - `input_source`: name or path of source of `input_lines`.
        - `initial_state`: name of initial state.
	*/
	this.runtimeInit();
	//if isinstance(input_lines, StringList):
	//self.input_lines = input_lines
	//else:
	//self.input_lines = StringList(input_lines, source=input_source)
	this.inputLines = inputLines;
	this.lineOffset = -1;
	this.currentState = initialState || this.initialState;
	let transitions = undefined;
	const results = [];
	let state = this.getState();
	let nextState;
	try {
	    [ context, result ] = state.bof(context);
	    results.push(result);
	    while(true) {
		try {
		    try {
			this.nextLine();
			[ context, nextState, result ] = this.checkLine(context, state, transitions);
		    }
		    catch(error) {
		    }
		}
		catch(error) {
		}
	    }
	}
	catch(error) {
	}
	this.observers = [];
	return results;
    }

    addStates(stateClasses) {
	// stub
    }

    runtimeInit() {
	Object.values(this.states).forEach(s => s.runtimeInit());
    }

    getState(nextState) {
	if(nextState) {
	    this.currentState = nextState;
	}
	
	    
}

export default StateMachine;
