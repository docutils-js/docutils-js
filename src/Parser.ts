import Component from './Component';
import { UnimplementedError } from './Exceptions';
import {Document, ParserArgs} from "./types";

abstract class Parser extends Component {
    protected debugFn: any;
    protected debug: boolean;

    public constructor(args: ParserArgs = {}) {
        super();
        this.componentType = 'parser';
        this.configSection = 'parsers';
        this.debug = args.debug || false;
        this.debugFn = args.debugFn;
    }

    /* istanbul ignore function */
    /* eslint-disable-next-line no-unused-vars */
    abstract parse(inputstring: string, document: Document): any | any[] | void;
    /* eslint-disable-next-line no-unused-vars */
    public setupParse(inputstring: string, document: Document): void {
    } ;
    abstract finishParse(): void;

}

export default Parser;
