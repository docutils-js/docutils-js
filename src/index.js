export const __version__ = '0.14js';
