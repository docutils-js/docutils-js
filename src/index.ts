/* eslint-disable-next-line import/prefer-default-export */
export { parse } from './parse';
export { StringOutput, StringInput } from './io';
export { StandaloneReader } from './readers/standalone';
export { newDocument } from './newDocument';
import * as nodes from './nodes';
import Writer from './Writer';
import Transform from './Transform';
export { Reader } from './Reader';
export { Publisher } from './Publisher'
export { defaults } from '../gen/defaults';
export { pojoTranslate } from './fn/pojoTranslate';
export { htmlTranslate } from './fn/htmlTranslate';
export { RSTParser } from './parsers/restructuredtext';
export { XMLWriter } from './writers/xml'

/* eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars */
const __docformat__ = 'reStructuredText';

export const __version__ = '0.14js';

export {
    nodes, Writer, Transform,
};
