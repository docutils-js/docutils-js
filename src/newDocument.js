import { document } from './nodes';
import { newReporter } from './utils';

export default function newDocument({ sourcePath }, settings) {
    const reporter = newReporter({ sourcePath }, settings);
    const attrs = {};
    if (typeof sourcePath !== 'undefined') {
        attrs.source = sourcePath;
    }

    const myDocument = new document(settings, reporter, '', [], attrs);
    myDocument.noteSource(sourcePath, -1);
    return myDocument;
}
