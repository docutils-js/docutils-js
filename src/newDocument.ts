import { document } from './nodes';
import newReporter from './newReporter';
import {Settings} from "../gen/Settings";
import { Document } from "./types";

/* eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars */
const __docformat__ = 'reStructuredText';

/**
 *  Return a new empty document object.
 *
 *  :Parameters:
 *      `source_path` : string
 *          The path to or description of the source text of the document.
 *      `settings` : optparse.Values object
 *          Runtime settings.  If none are provided, a default core set will
 *          be used.  If you will use the document object with any Docutils
 *          components, you must provide their default settings as well.  For
 *          example, if parsing, at least provide the parser settings,
 *          obtainable as follows::
 *
 *              settings = docutils.frontend.OptionParser(
 *                  components=(docutils.parsers.rst.Parser,)
 *                  ).get_default_values()
 * {@link module:nodes~document}
 * @returns {module:nodes~document} New document
 * @see module:newDocument~newDocument
 */
function newDocument(args: { sourcePath: string }, settings: Settings): Document{
    const {sourcePath} = args;
    const reporter = newReporter({ sourcePath }, settings);
    const attrs: { source?: string }= {};
    if (typeof sourcePath !== 'undefined') {
        attrs.source = sourcePath;
    }

    // eslint-disable-next-line new-cap
    const myDocument = new document(settings, reporter, '', [], attrs);
    myDocument.noteSource(sourcePath, -1);
    return myDocument;
}

export { newDocument };

export default newDocument;
