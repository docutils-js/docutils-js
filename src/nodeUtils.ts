// ========================================
import { NodeInterface } from "./types";
import xmlescape from "xml-escape";
import { Text } from "./nodes";

/**
 * convert a node to XML
 */
export function nodeToXml(node: NodeInterface): string {
    if (node instanceof Text) {
        const text = xmlescape(node.astext());
        return text;
    }
    if (node.hasChildren()) {
        return [node.starttag(), ...node.getChildren().map((c: NodeInterface): string => nodeToXml(c)), node.endtag()].join("");
    }
    return node.emptytag();
}
/**
 * Return a whitespace-normalized name.
 */
export function whitespaceNormalizeName(name: string): string {
    return name.replace(/\s+/, " ");
}

export function fullyNormalizeName(name: string): string {
    return name.toLowerCase().replace(/\s+/, " ");
}
