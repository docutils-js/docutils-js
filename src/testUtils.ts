import { newDocument } from './newDocument';
import newReporter    from './newReporter';
import { defaultPublisherOptions } from '../src/constants';
import { getDefaultSettings } from './settingsHelper';
import winston, { format, transports, LoggerOptions, Logger } from 'winston';
import { defaultConsoleLogLevel } from './constants';
import {Publisher} from './Publisher';
import path from 'path';
import StateFactory from './parsers/rst/StateFactory';
import { LoggerType } from './types';
import process from 'process';

let logger: LoggerType|undefined;

export function createLogger(options?: LoggerOptions): Logger {
    if(logger !== undefined) {
        return logger;
    }
    let myOpt = options === undefined ? {} : { ...options};
    if(myOpt.format === undefined) {
        myOpt.format = format.json();
    }
    if(myOpt.transports === undefined) {
        myOpt.transports = [
            new transports.Console({level: defaultConsoleLogLevel}),
            new transports.File({filename: `${path.basename(process.argv[1])}-${process.pid}.log`, level: 'silly'}),
        ];
    }
    logger = winston.createLogger(myOpt);
    return logger;
}

export function createPublisher() {
    const publisher = new Publisher({ ...defaultPublisherOptions, logger: createLogger() });
    publisher.setComponents(defaultPublisherOptions.readerName, defaultPublisherOptions.parserName, 'xml');
    return publisher;
}

export function createStateMachine() {
}

export function createStateFactory(): StateFactory {
return new StateFactory({ logger: createLogger() });
}

export function createNewDocument(sourcePath: string='default') {
    return newDocument({sourcePath, logger: createLogger()}, getDefaultSettings());
}
  
export function createNewReporter(sourcePath: string='default') {
    return newReporter({sourcePath}, getDefaultSettings());
}
