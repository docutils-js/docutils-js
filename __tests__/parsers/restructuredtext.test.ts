import * as fs from "fs";
import * as path from "path";
import Parser from "../../src/parsers/restructuredtext";
import newDocument from "../../src/newDocument";
import {nodeToXml} from "../../src/nodeUtils";
import{ getDefaultSettings} from'../../src/';
import{ createNewDocument, createLogger} from'../../src/testUtils';

const ReadmeRst = fs.readFileSync(path.join(__dirname, '../../README.rst'), { encoding: 'UTF-8' });

test.only('1', () => {
const logger = createLogger();
    const p = new Parser({ logger});
    const document = createNewDocument();
    p.parse('* a bullet point', document);
    expect(nodeToXml(document)).toMatchSnapshot();
});

test('rst parser no input', () => {
const logger = createLogger();
    const p = new Parser({logger});
    const document = createNewDocument();
    expect(() => p.parse('', document)).toThrow();
});

test('readme rst', () => {
const logger = createLogger();
    const p = new Parser({logger});
    const document = createNewDocument();
    p.parse(ReadmeRst, document);
    expect(nodeToXml(document)).toMatchSnapshot();
});
