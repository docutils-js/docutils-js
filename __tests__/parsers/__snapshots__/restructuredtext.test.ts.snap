// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1 1`] = `"<document source=\\"default\\"><bullet_list bullet=\\"*\\"><list_item><paragraph>a bullet point</paragraph></list_item></bullet_list></document>"`;
